feat: Тандердом для призраков

Vladisvell opened the pull request:
<!-- Пишите **НИЖЕ** заголовков и **ВЫШЕ** комментариев, иначе ваш текст может не отобразиться. -->
<!-- В Contributing.MD вы можете найти некоторые рекомендации к оформлению пулл-реквеста. -->

## Описание
<!-- Опишите, что делает ваш ПР. Документировать каждую деталь не требуется, просто укажите основные изменения. -->
Обновление, добавляющее Тандердом для призраков.
**!ВАЖНО! Добавлен Preference для роли Thunderdome, отметьте её для того, чтобы играть на арене.**

- **Стать участником Тандердома может каждый, кто отметит в своих Game Preferences роль **_Be Thunderdome_**.**
- Длительность схватки: 5 минут.
- Максимальное число игроков на тандердоме: 16 человек.
- Через mob spawner в меню призрака можно начать голосование за начало боя на арене
- Доступны три режима: Ближний, Дальний, Смешанный.
- В ближнем бою игрок будет иметь возможность драться с использованием боевых техник и одного случайного предмета для стычек на короткой дистанции.
- В дальнем бою игрок получит оружие для ведения дистанционных схваток.
- В смешанном режиме будет и то и другое.
- Все получают имплант адренала для более динамичных сражений.
- Добавлен имплант, убивающий участников Тандердома спустя определенное время (5 минут).
- Вы можете выйти с арены, просто нажав на кнопку активации импланта смерти (в виде иконки костей).
- В зависимости от режима игрокам будет доступна разная площадь арены.

Техническая информация описана в коммите https://github.com/ss220-space/Paradise/pull/2912/commits/345a55ea0b8826b737356610cc3c5b3a0a97cafc

- Начало и конец тандердома логируются
- Пул оружия не содержит телепортационные вещи
- Арена неразрушаема
- Освещение постоянное
- Расположена в админ-зоне
- Возможно расширение арены в будущем
- У участников Тандердома теперь есть рюкзаки с около-бесконечной ёмкостью
- Отсутствуют наушники, радиосвязь.
- Очищение арены от всех объектов кроме поллера, пола, шаттеров, стен спустя 5 минут после начала схватки.
- Телепортирующие вещи не работают (ввиду особенности админ-зоны)

Для мапперов:
- Добавлен thunderdome_poller, являющийся центром арены, можно использовать как моб_спавн.
- Можно инициализировать собственный экземпляр датума тандердома и настраивать по своему вкусу.


Для кодеров:
- Для ```/datum/spawners_menu/ui_act``` изменён принцип, по которому можно определить валидность спавнера, теперь можно использовать переопределенный метод ```is_mob_spawnable```, возвращающий TRUE, чтобы задать желаемый объект, как возможный к использованию через ```mob spawn menu```
- Добавлены дефайны для тестирования

## Ссылка на предложение/Причина создания ПР
<!-- Здесь оставьте ссылку на сообщение в #отчеты-по-предложениям, чтобы подтвердить, что ваше предложение одобрено. Либо укажите, почему этот ПР должен пройти без предложки. -->
<!-- Пример ссылки: https://discord.com/channels/617003227182792704/755125334097133628/ID-сообщения -->
https://discord.com/channels/617003227182792704/755125334097133628/1107911273351491645

## Демонстрация изменений
<!-- Здесь вы можете показать изменения внешне, к примеру новые спрайты, звуки или изменения карты. Или написать, что именно изменилось для игроков. Этот пункт полностью опционален и его можно удалить. -->
Игрокам будет доступна площадь:
- 13x13 для ближнего боя.
- 21х21 для дальнего и смешанного боя.
![изображение](https://github.com/ss220-space/Paradise/assets/73733747/551f9200-908a-48d0-a13f-cc2f710ec9f9)


[Демонстрация голосования и спауна.](https://media.discordapp.net/attachments/734823601110515882/1115541003844198420/37.gif?width=1363&height=676)

🆑Vladisvell
feat: Добавлен Тандердом для обычных игроков c доступом через Mob Spawners.
feat: Добавлена возможность призракам со включенным Preference на Thunderdome участвовать в аренах
feat: Три режима Тандердома (ближний, дальний, смешанный)
feat: Случайный выбор оружия для трёх режимов Тандердома
ref: Теперь можно использовать в моб спавнерах все, что возвращает TRUE при вызове is_mob_spawnable
/🆑

Vakareet committed:
Create thunderdome cleaner device icon

Vakareet committed:
Add thunderdome cleaner device

Vakareet committed:
Add new thunderdome area

Vakareet committed:
Create new thunderdome arena on admin Z-level

Vakareet committed:
Add shutters for different thunderdome modes

Vakareet committed:
Add tdome CQC area

Vakareet committed:
Create tdomeCQC area inside of the shutters on the New Thunderdome Arena

Vakareet committed:
Unfinished files for thunderdome poll

Vladisvell committed:
Thunderdome for ghosts large commit
- Changed Admin_Zone.dmm, added thunderdome pollers for ranged, melee, mixed fights
- Added role preference for Thunderdome player
- Changed how spawners menu work, now you neeed to override is_mob_spawnable() method to return TRUE if you want to use it with mob spawners menu from ghost
- Removed thunderdome_purifier.dm due it's redundancy.
- Added random items giving to ranged, melee, mixed fighters.
- Overriden mobspawners for thunderdome mob_spawn in order to snap player into doll right after poll vote.
- Added to mob_spawn tts boolean which determines if player can choose tts seed.
- Added postponed death implant, which will kill you after some time passed.
- Added three outfits and thunderdome classes mob_spawns.
- Added associative list picker which allows random picking keys with values from a list.
- Area will be cleansed after DEFAULT_TIME_LIMIT or defined otherwise in thunderdome_battle datum.
- Changed thunderdome previews icons.
- Added thunderdome_battle.dm to paradise.dme.
- Now shutters work according to state of an arena (ranged, melee, mixed)

Vladisvell committed:
Changed Admin_Zone.dmm to tgm format

Vladisvell committed:
- Added new comments
- Changed challengers' outfit to durathread
- Now they wear HoS head helmet
- Changed stunbaton in random poll of melee items to loaded baton
- Reduced number of sabers to 1
- Typo fixes

Den109G commented on code/modules/thunderdome/thunderdome_battle.dm:
```diff
@@ -0,0 +1,366 @@
+		if(istype(A, /obj/machinery/door/poddoor) || istype(A, /obj/thunderdome_poller))
+			continue
+		qdel(A)
+	}
```
`{` и `}` для `for` выглядят странно, учитывая что у нас не пишут их (да и выше в `get_random_items` не написал).
Просто приведение к общему стилю.

Vladisvell replied:
Спасибо, исправил

Vladisvell committed:
code style fix

Vladisvell committed:
Removed energy katana from melee pool

Vladisvell committed:
removed admin_zone.dmm from includes of paradise.dme

Vladisvell committed:
linter issue fix attempt 1

Rerik007 commented on code/__DEFINES/role_preferences.dm:
```diff
@@ -64,6 +65,7 @@ GLOBAL_LIST_INIT(special_roles, list(
 	ROLE_ELITE,											// Lavaland Elite
 	ROLE_MORPH, 										// Morph
 	ROLE_OPERATIVE = /datum/game_mode/nuclear, 			// Operative
+	ROLE_THUNDERDOME,									// Thunderdome
```
Отсюда строится лист в лобби, остальное по алфавиту отсортировано, нельзя ли и это тоже переместить вниз, после вора?

Vladisvell replied:
Хорошо, исправил

Rerik007 commented on code/modules/awaymissions/corpse.dm:
```diff
@@ -36,6 +36,7 @@
 	var/allow_species_pick = FALSE
 	var/allow_gender_pick = FALSE
 	var/allow_name_pick = FALSE
+	var/allow_tts_pick = FALSE
```
Хм, теперь за свармера не удастся выбрать голос? Вряд ли кто-то заметит конечно..

Vladisvell replied:
Ладно...поставил на TRUE

Rerik007 commented on code/modules/thunderdome/thunderdome_battle.dm:
```diff
@@ -0,0 +1,367 @@
+#define RANGED_ARENA_RADIUS 10
+#define VOTING_POLL_TIME 60 SECONDS
+#define MAX_PLAYERS_COUNT 16
+#define MIN_PLAYERS_COUNT 1
```
Не стоит ли 2 поставить? Я бы не хотел на 5 минут оставаться там с единственной возможностью выбраться только забив себя. Или хотя бы разрешить игрокам использовать тот имплант 5-ти минутный, чтобы после победы можно было нормально выйти.

Vladisvell replied:
Верно, спасибо за замечание

Vladisvell replied:
> Вот это админзона поменялась конечно... А если серьёзно то в основном я бы хотел знать, а что делать тем, кто уже победил или появился один. Нельзя ли им дать другой быстрый и понятный способ выйти в таких случаях?

Добавил импланту смерти кнопку, которой можно будет мгновенно выйти в госты

Rerik007 reviewed (commented):
Вот это админзона поменялась конечно... А если серьёзно то в основном я бы хотел знать, а что делать тем, кто уже победил или появился один. Нельзя ли им дать другой быстрый и понятный способ выйти в таких случаях?

Vladisvell committed:
фиксы по ревью

Vladisvell committed:
tgm convertion reversion, now should be adequate

Vladisvell committed:
map fix attempt 2

Vladisvell committed:
another try fixing map issue

Vladisvell committed:
Merge branch 'master220' into thunderdome_for_ghosts

Vladisvell committed:
-Added logging (and more)
-Added infinite backpack for thunderdome fighters
-Added check for eligible(lity?) for polling
-Added notification for ghosts when thunderdome is ready to poll again
-Added notification for ghosts when thunderdome can't be started due to low player count willing to play
